ui/raidboss/data/03-hw/dungeon/gubal_library_hard.js

```diff
@@ -0,0 +1,300 @@
+'use strict';
+
+// The Great Gubal Library--Hard
+[{
+  zoneRegex: /Great Gubal Library \(Hard\)/,
+  timelineFile: 'gubal_library_hard.txt',
+  timelineTriggers: [
+    {
+      id: 'Gubal Hard Triclip',
+      regex: /Triclip/,
+      beforeSeconds: 5,
+      condition: function(data) {
+        return data.role == 'healer' || data.role == 'tank';
+      },
+      infoText: {
+        en: 'Tank Buster',
+      },
+    },
+    {
+      id: 'Gubal Hard Searing Wind',
+      regex: /Searing Wind/,
+      beforeSeconds: 5,
+      infoText: function(data) {
+        if (data.role == 'tank') {
+          return {
+            en: 'Tank Cleave on you',
+          };
+        }
+        return {
+          en: 'Avoid tank cleave',
+        };
+      },
+    },
+    {
+      id: 'Gubal Hard Properties of Darkness',
+      regex: /Darkness \(buster\)/,
+      beforeSeconds: 5,
+      condition: function(data) {
+        return data.role == 'healer' || data.role == 'tank';
+      },
+      infoText: {
+        en: 'Tank Buster',
+      },
+    },
+  ],
+  triggers: [
+    {
+      id: 'Gubal Hard Bibliocide',
+      regex: / 14:1945:Liquid Flame starts using Bibliocide/,
+      condition: function(data) {
+        return data.role == 'healer';
+      },
+      infoText: {
+        en: 'AoE',
+      },
+    },
+    {
+      // Ferrofluid handling is shamelessly copied from Hades Normal
+      id: 'Gubal Hard Ferrofluid Collect',
+      regex: / 1B:\y{ObjectId}:(\y{Name}):....:....:(0030|0031):/,
+      run: function(data, matches) {
+        data.fluid = data.fluid || {};
+        if (matches[2] == '0030')
+          data.fluid[matches[1]] = 'positive';
+        if (matches[2] == '0031')
+          data.fluid[matches[1]] = 'negative';
+      },
+    },
+    {
+      id: 'Gubal Hard Ferrofluid Execute',
+      regex: / 1B:\y{ObjectId}:\y{Name}:....:....:(?:0030|0031):/,
+      suppressSeconds: 5,
+      delaySeconds: 0.5,
+      infoText: function(data) {
+        if (data.fluid[data.me] == data.fluid['Liquid Flame']) {
```

Comment:
Ah, this is tricky because of the translation.  You can tell from the ObjectId whether something is an enemy or not, because it starts with '4'.

Does the boss always get listed first before the player or vice versa?

I'm mostly curious if this could be simplified somehow.

Reply:
It looked like it does? Or at least it seemed to the times I ran this. I was hesitant to rely on relative timing since I know I've seen encounters where that wasn't the case, hence the handling via collection and comparison. Do you have any suggestions on where similar simpler patterns might already exist in the repository?

Reply:
I've depended on that for other things, but I think you're right that it feels like it might not be that robust.

I think a simpler trigger might look like this.  A single trigger that does this:
* condition: on me or the boss (by object id comparison)
* preRun: store the headmarker in a list (and create the list if it doesn't exist)
* infoText: if there aren't two things in the list, return.  If there are two things in the list, compare them and print an alert.  Also, delete the list.

Reply:
Oh, I see where you're going with that. You're right, that would break in non-English locales. Okay. So maybe change over the collection regex to capture the ObjectID instead, then filter for which one starts with 4 and set that one as the primary. Something like this but better...

```
    {
      id: 'Gubal Hard Ferrofluid Collect',
      regex: / 1B:\(y{ObjectId}):\y{Name}:....:....:(0030|0031):/,
      run: function(data, matches) {
        if (matches[1].slice(0,1) == '4') {
          data.flameId = data.flameId || matches[1];
        }
        data.fluid = data.fluid || {};
        if (matches[2] == '0030')
          data.fluid[matches[1]] = 'positive';
        if (matches[2] == '0031')
          data.fluid[matches[1]] = 'negative';
      },
    },
    {
      id: 'Gubal Hard Ferrofluid Execute',
      regex: / 1B:(\y{ObjectId}):(\y{Name}):....:....:(?:0030|0031):/,
      delaySeconds: 0.5,
      condition: function(data, matches) {
        return matches[2] == data.me;
      }
      infoText: function(data) {
        if (data.fluid[matches[1]] == data.fluid[data.flameId]) {
          return {
            en: 'Close to boss',
          };
        }
        return {
          en: 'Away from boss',
        };
      },
    },
```

Reply:
Hmm, I need to refresh the page before commenting, it seems. Yes, your suggestion would definitely be easier than what I have there. I'll poke at that tomorrow.

Reply:
Well, never mind, it's *technically* tomorrow!

```
    {
      id: 'Gubal Hard Ferrofluid',
      regex: / 1B:(\y{ObjectId}):(\y{Name}):....:....:(0030|0031):/,
      condition: function(data, matches) {
        return data.me == matches[2] || matches[1].slice(0,1) == '4';
      },
      preRun: function(data, matches) {
        data.markers = data.markers || [];
        data.markers.push(matches[3]);
      },
      infoText: function(data) {
        if (data.markers.length < 2)
          return;
        if (data.markers[1] == data.markers[2]) {
          delete data.markers;
          return {
            en: 'Close to boss',
          };
        }
        delete data.markers;
        return {
          en: 'Away from boss',
        };
      },
    },
```

I appreciate the pointers on this.  It's definitely better than stacking three triggers and is locale-agnostic too. I don't like duplicating the deletion, but adding a `run` command would end up being just as long. Any further suggestions on it?

Reply:
Hmm! You could store the markers off into a separate variable, and then delete them?
```
let sameMarkers = data.markers[0] == data.markers[1];
delete data.markers;
if (sameMarkers) {
// etc
```

Reply:
Definitely better. It is done!